modules/navier_stokes/src/executioners/PIMPLESolve.C

```diff
@@ -0,0 +1,456 @@
+#include "LinearSystem.h"
+
+using namespace libMesh;
+
+InputParameters
+PIMPLESolve::validParams()
+{
+  InputParameters params = SIMPLESolveBase::validParams();
+  params.addParam<unsigned int>(
+      "num_piso_iterations",
+      0,
+      "The number of PISO iterations without recomputing the momentum matrix.");
+
+  return params;
+}
+
+PIMPLESolve::PIMPLESolve(Executioner & ex)
+  : SIMPLESolveBase(ex),
+    _pressure_sys_number(_problem.linearSysNum(getParam<SolverSystemName>("pressure_system"))),
+    _pressure_system(_problem.getLinearSystem(_pressure_sys_number)),
+    _energy_sys_number(_has_energy_system
+                           ? _problem.linearSysNum(getParam<SolverSystemName>("energy_system"))
+                           : libMesh::invalid_uint),
+    _energy_system(_has_energy_system ? &_problem.getLinearSystem(_energy_sys_number) : nullptr),
+    _num_piso_iterations(getParam<unsigned int>("num_piso_iterations"))
+
+{
+
+  // We fetch the systems and their numbers for the momentum equations.
+  for (auto system_i : index_range(_momentum_system_names))
+  {
+    _momentum_system_numbers.push_back(_problem.linearSysNum(_momentum_system_names[system_i]));
+    _momentum_systems.push_back(&_problem.getLinearSystem(_momentum_system_numbers[system_i]));
+    _systems_to_solve.push_back(_momentum_systems.back());
+  }
+
+  _systems_to_solve.push_back(&_pressure_system);
+
+  if (_has_energy_system)
+    _systems_to_solve.push_back(_energy_system);
+
+  // and for the passive scalar equations
+  if (_has_passive_scalar_systems)
+    for (auto system_i : index_range(_passive_scalar_system_names))
+    {
+      _passive_scalar_system_numbers.push_back(
+          _problem.linearSysNum(_passive_scalar_system_names[system_i]));
+      _passive_scalar_systems.push_back(
+          &_problem.getLinearSystem(_passive_scalar_system_numbers[system_i]));
+      _systems_to_solve.push_back(_passive_scalar_systems.back());
+    }
+}
+
+void
+PIMPLESolve::linkRhieChowUserObject()
+{
+  _rc_uo =
+      const_cast<RhieChowMassFlux *>(&getUserObject<RhieChowMassFlux>("rhie_chow_user_object"));
+  _rc_uo->linkMomentumPressureSystems(
+      _momentum_systems, _pressure_system, _momentum_system_numbers);
+
+  // Initialize the face velocities in the RC object
+  if (!_app.isRecovering())
+    _rc_uo->initFaceMassFlux();
+  _rc_uo->initCouplingField();
+}
+
+std::vector<std::pair<unsigned int, Real>>
+PIMPLESolve::solveMomentumPredictor()
+{
+  // Temporary storage for the (flux-normalized) residuals from
+  // different momentum components
+  std::vector<std::pair<unsigned int, Real>> its_normalized_residuals;
+
+  LinearImplicitSystem & momentum_system_0 =
+      libMesh::cast_ref<LinearImplicitSystem &>(_momentum_systems[0]->system());
+
+  PetscLinearSolver<Real> & momentum_solver =
+      libMesh::cast_ref<PetscLinearSolver<Real> &>(*momentum_system_0.get_linear_solver());
+
+  momentum_solver.reuse_preconditioner(true);
```

Comment:
where is the step to "refresh" it on the next time step?

Reply:
Great catch! I will add the preconditioner refresh after every simple iteration. At some point I'll add an adaptive preconditioner refresh as well (so when residuals are low we don't refresh anymore).